If association is a hash-like object preloading fails

bpohoriletz opened the pull request:
### If association is a hash-like object preloading fails

If you pass a hash-like object to preload associations (for example ActionController::Parameters)
preloader will fail with the ArgumentError.
This change allows passing objects that may be converted to a Hash or String into a preloader

### Other Information

` bundle exec ruby -Itest test/cases/associations/eager_test.rb` passes

composerinteralia commented on activerecord/lib/active_record/associations/preloader.rb:
```diff
@@ -99,13 +99,13 @@ def preload(records, associations, preload_scope = nil)
-            preloaders_for_one(association, records, scope)
-          when String
+          case
+          when true == association.respond_to?(:to_hash)
```
I think we should get rid of the `true ==` and use `if/elsif`

composerinteralia commented on activerecord/test/cases/associations/eager_test.rb:
```diff
@@ -1511,6 +1511,49 @@ def test_preloading_has_many_through_with_custom_scope
 
+  # Associations::Preloader#preloaders_on works with hash-like objects
+  test "preloading works when object that responds to :to_hash is passed" do
+    post_id = Comment.where(author_id: nil).where.not(post_id: nil).first.post_id
```
Does the specific `post_id` matter in these tests? Can we use `Post.preload(...).first` instead?

composerinteralia commented on activerecord/test/cases/associations/eager_test.rb:
```diff
@@ -1511,6 +1511,49 @@ def test_preloading_has_many_through_with_custom_scope
   end
 
+  # Associations::Preloader#preloaders_on works with hash-like objects
+  test "preloading works when object that responds to :to_hash is passed" do
```
Could avoid the passive voice here and write `"preloading works with object that responds to :to_hash"`

bpohoriletz commented:
`bundle exec rake test:sqlite3` passes

composerinteralia commented on activerecord/test/cases/associations/eager_test.rb:
```diff
@@ -1511,6 +1511,49 @@ def test_preloading_has_many_through_with_custom_scope
+  # Associations::Preloader#preloaders_on works with symbol-like objects
+  test "preloading works when object that responds to :to_sym is passed" do
+    post_id = Comment.where(author_id: nil).where.not(post_id: nil).first.post_id
+    custom_symbol = (1..10)
```
If this specific range is not important to the test, you can use `Object.new`. You could also just use a string or some other object that already responds to `to_sym`.

composerinteralia reviewed (changes requested):
Thanks for the PR! This may be making the method more flexible than it needs to be, but  I suppose if an object is hash-like or string-like enough to respond to `to_hash` or `to_str` you might expect it to work here.

Could you also squash this into one commit?

~~Don't worry about the Travis failures on Ruby 2.5.0 `av:ujs` and `railties`. There is a Travis bug that has caused them to fail for a while.~~ Should be fixed now

Commit:
If association is a hash-like object preloading fails

If you pass a hash-like object to preload associations (for example ActionController::Parameters)
preloader will fail with the ArgumentError.
This change allows passing objects that may be converted to a Hash or String into a preloader

bpohoriletz commented:
thank you for the review @composerinteralia, I've made requested changes

composerinteralia commented:
@bpohoriletz it might be worth rebasing this on `master` so you can get a passing travis build.
@eileencodes I left an initial review, but this still needs review by a committer.

bpohoriletz commented:
@composerinteralia thank you for the advise, done now

oniofchaos commented:
@bpohoriletz looks like this just fell through the cracks, can you fix up the merge conflict?

bpohoriletz commented:
@oniofchaos done

oniofchaos commented:
@eileencodes would you be able to give this a review? Thanks 💝

eileencodes commented:
@bpohoriletz thanks for rebasing.

@oniofchaos I know you mean well, but please don't ping reviewers - I already got one email that Bohdan finished rebasing. Thanks for understanding! ❤️

faucct commented on activerecord/test/cases/associations/eager_test.rb:
```diff
@@ -1618,6 +1618,32 @@ def test_preloading_has_many_through_with_custom_scope
+    end
+    assert_equal("1..10 was not recognized for preload", exception.message)
+  end
+
```
These aren't the cases you are fixing: the tests work just fine without the changes.
